src/math/dd/dd_bdd.cpp

```diff
@@ -1014,14 +996,133 @@ namespace dd {
     bdd bdd_manager::mk_ult(bddv const& a, bddv const& b) { return mk_ule(a, b) && !mk_eq(a, b); }
     bdd bdd_manager::mk_ugt(bddv const& a, bddv const& b) { return mk_ult(b, a); }
 
-    bdd_manager::bddv bdd_manager::mk_add(bddv const& a, bddv const& b) {
+    bddv bdd_manager::mk_add(bddv const& a, bddv const& b) {
         SASSERT(a.size() == b.size());
         bdd carry = mk_false();
-        bddv result;
+        bddv result(this);
+#if 0
         for (unsigned i = 0; i < a.size(); ++i) {
+            result.push_back(carry ^ a[i] ^ b[i]);
             carry = (carry && a[i]) || (carry && b[i]) || (a[i] && b[i]);
+        }
+#else
+        if (a.size() > 0) {
+            result.push_back(a[0] ^ b[0]);
+        }
+        for (unsigned i = 1; i < a.size(); ++i) {
+            carry = (carry && a[i-1]) || (carry && b[i-1]) || (a[i-1] && b[i-1]);
             result.push_back(carry ^ a[i] ^ b[i]);
         }
+#endif
+        return result;
+    }
+
+    void bdd_manager::bddv_shl(bddv &a) {
+        for (unsigned j = a.size(); j-- > 1; ) {
+            a[j] = a[j - 1];
+        }
+        a[0] = mk_false();
+    }
+
+    bddv bdd_manager::mk_mul(bddv const& a, bddv const& b) {
+        SASSERT(a.size() == b.size());
+        bddv a_shifted = a;
+        bddv result = mk_zero(a.size());
+        for (unsigned i = 0; i < b.size(); ++i) {
+#if 1
+            bddv s = a_shifted;
+            for (unsigned j = i; j < b.size(); ++j) {
+                s[j] &= b[i];
+            }
+            result = mk_add(result, s);
+#else
+            // From BuDDy's bvec_mul. It seems to compute more intermediate BDDs than the version above?
```

Comment:
not sure it is worth spending time on general bdd multiplication. It is too expensive to be used upstream for anything with larger bit-widths.

Reply:
Ok; I was just wondering why BuDDy is using a seemingly more complicated version. In any case, both versions seem to work correctly.

Reply:
We can compare several sources for encoding multipliers. Also CUDD. In the end the BDDs are unique modulo ordering and there is never a good ordering. Just stay with the simple grade school encoding.